webhook/server/server.go

```diff
@@ -0,0 +1,437 @@
+		err = fmt.Errorf("received object is not an AdmissionReview")
+	}
+
+	return ar, err
+}
+
+func deserializeNetworkAttachmentDefinition(ar *v1beta1.AdmissionReview) (types.NetworkAttachmentDefinition, error) {
+	/* unmarshal NetworkAttachmentDefinition from AdmissionReview request */
+	netAttachDef := types.NetworkAttachmentDefinition{}
+	err := json.Unmarshal(ar.Request.Object.Raw, &netAttachDef)
+	return netAttachDef, err
+}
+
+func deserializePod(ar *v1beta1.AdmissionReview) (v1.Pod, error) {
+	/* unmarshal Pod from AdmissionReview request */
+	pod := v1.Pod{}
+	err := json.Unmarshal(ar.Request.Object.Raw, &pod)
+	return pod, err
+}
+
+func handleValidationError(w http.ResponseWriter, ar *v1beta1.AdmissionReview, orgErr error) {
+	err := prepareAdmissionReviewResponse(false, orgErr.Error(), ar)
+	if err != nil {
+		err := logging.Errorf("error preparing AdmissionResponse: %s", err.Error())
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	writeResponse(w, ar)
+}
+
+func writeResponse(w http.ResponseWriter, ar *v1beta1.AdmissionReview) {
+	logging.Debugf("sending response to the Kubernetes API server")
+	resp, _ := json.Marshal(ar)
+	w.Write(resp)
+}
+
+func validateHandler(w http.ResponseWriter, req *http.Request) {
+	/* read AdmissionReview from the HTTP request */
+	ar, httpStatus, err := readAdmissionReview(req)
+	if err != nil {
+		http.Error(w, err.Error(), httpStatus)
+		return
+	}
+
+	netAttachDef, err := deserializeNetworkAttachmentDefinition(ar)
+	if err != nil {
+		handleValidationError(w, ar, err)
+		return
+	}
+
+	/* perform actual object validation */
+	allowed, err := validateNetworkAttachmentDefinition(netAttachDef)
+	if err != nil {
+		handleValidationError(w, ar, err)
+		return
+	}
+
+	/* perpare response and send it back to the API server */
+	err = prepareAdmissionReviewResponse(allowed, "", ar)
+	if err != nil {
+		logging.Errorf(err.Error())
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	writeResponse(w, ar)
+}
+
+func getNetworkAttachmentDefinition(namespace, name string) (*types.NetworkAttachmentDefinition, error) {
+	path := fmt.Sprintf("/apis/k8s.cni.cncf.io/v1/namespaces/%s/network-attachment-definitions/%s", namespace, name)
+	rawNetworkAttachmentDefinition, err := clientset.ExtensionsV1beta1().RESTClient().Get().AbsPath(path).DoRaw()
+	if err != nil {
+		err := logging.Errorf("could not get Network Attachment Defintion %s/%s: %s", namespace, name, err)
+		return nil, err
+	}
+
+	networkAttachmentDefinition := types.NetworkAttachmentDefinition{}
+	json.Unmarshal(rawNetworkAttachmentDefinition, &networkAttachmentDefinition)
+
+	return &networkAttachmentDefinition, nil
+}
+
+func parsePodNetworkSelectionElement(selection string) (*types.NetworkSelectionElement, error) {
+	var namespace, name, netInterface string
+	var networkSelectionElement *types.NetworkSelectionElement
+
+	units := strings.Split(selection, "/")
+	switch len(units) {
+	case 1:
+		namespace = "default"
```

Comment:
This violates CRD document. If namespace is not specified then the pods namespace, not the default namespace, should be assumed.

See here: https://docs.google.com/document/d/1Ny03h6IDVy_e_vmElOqR7UdTPAG_RNydhVE1Kx54kFQ/edit#

"4.1.1 Comma-delimited Format
This format is intended to be a very simple, user-friendly format, composed only of NetworkAttachmentDefinition object references separated by commas.  The format of each object reference is either (a) <NetworkAttachmentDefinition object name> to reference a NetworkAttachmentDefinition object in the pod’s namespace, or (b) <NetworkAttachmentDefinition object namespace>/<NetworkAttachmentDefinition object name> to reference a NetworkAttachmentDefinition object in a different namespace."

Reply:
fixed